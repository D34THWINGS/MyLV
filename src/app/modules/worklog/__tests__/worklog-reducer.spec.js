import worklogReducer from '../worklog-reducer'
import {
  WORKLOG_EMPTY_DAY,
  WORKLOG_FILL_AFTERNOON,
  WORKLOG_FILL_DAY,
  WORKLOG_FILL_MONTH,
  WORKLOG_FILL_MORNING,
  WORKLOG_FILL_WEEK,
  WORKLOG_SAVE_SUCCESS
} from '../worklog-actions'

jest.unmock('../worklog-reducer')
jest.unmock('../worklog-actions')

jest.mock('../calendar-constants')
jest.mock('moment', () => jest.fn((stringDate) => {
  let counter = 0
  const momentMock = {
    startOf: jest.fn(() => momentMock),
    subtract: jest.fn(() => momentMock),
    add: jest.fn(() => {
      counter += 1
      return momentMock
    }),
    format: jest.fn(() => `foo-${counter}`),
    daysInMonth: jest.fn(() => 30),
    day: jest.fn(() => Number(stringDate.slice(-2)) % 7)
  }
  return momentMock
}))

describe('worklog/reducer', () => {
  describe('WORKLOG_FILL_MORNING', () => {
    it('should fill morning entry with given label', () => {
      // Given
      const action = { type: WORKLOG_FILL_MORNING, payload: { date: 'hello', label: 'world' } }

      // When
      const state = worklogReducer({ entries: { foo: 'bar' } }, action)

      // Then
      expect(state).toEqual({
        entries: {
          foo: 'bar',
          'hello-am': 'world'
        },
        pending: {
          'hello-am': 'world'
        }
      })
    })
  })

  describe('WORKLOG_FILL_AFTERNOON', () => {
    it('should fill afternoon entry with given label', () => {
      // Given
      const action = { type: WORKLOG_FILL_AFTERNOON, payload: { date: 'hello', label: 'world' } }

      // When
      const state = worklogReducer({ entries: {} }, action)

      // Then
      expect(state).toEqual({ entries: { 'hello-pm': 'world' }, pending: { 'hello-pm': 'world' } })
    })
  })

  describe('WORKLOG_FILL_DAY', () => {
    it('should fill entry with given label', () => {
      // Given
      const action = { type: WORKLOG_FILL_DAY, payload: { day: 'hello', label: 'world' } }

      // When
      const state = worklogReducer({ entries: {} }, action)

      // Then
      expect(state).toEqual({
        entries: { 'hello-am': 'world', 'hello-pm': 'world' },
        pending: { 'hello-am': 'world', 'hello-pm': 'world' }
      })
    })
  })

  describe('WORKLOG_FILL_WEEK', () => {
    it('should fill an entry for each day of the week with given label', () => {
      // Given
      const action = { type: WORKLOG_FILL_WEEK, payload: { day: 'hello', label: 'world' } }

      // When
      const state = worklogReducer({ entries: {} }, action)

      // Then
      expect(state).toMatchSnapshot()
    })
  })

  describe('WORKLOG_FILL_MONTH', () => {
    it('should fill an entry for each working day of the month with given label', () => {
      // Given
      const action = { type: WORKLOG_FILL_MONTH, payload: { month: '0000-00', label: 'world' } }

      // When
      const state = worklogReducer({ entries: {} }, action)

      // Then
      expect(state).toMatchSnapshot()
    })
  })

  describe('WORKLOG_EMPTY_DAY', () => {
    it('should empty given entry date', () => {
      // Given
      const action = { type: WORKLOG_EMPTY_DAY, payload: { day: 'foo' } }

      // When
      const state = worklogReducer({ entries: { 'foo-am': 'bar', 'foo-pm': 'bar' } }, action)

      // Then
      expect(state).toEqual({
        entries: { 'foo-am': undefined, 'foo-pm': undefined },
        pending: { 'foo-am': null, 'foo-pm': null }
      })
    })
  })

  describe('WORKLOG_SAVE_SUCCESS', () => {
    it('should empty pending changes', () => {
      // Given
      const action = { type: WORKLOG_SAVE_SUCCESS }

      // When
      const state = worklogReducer({ pending: { 'foo-am': 'bar', 'foo-pm': 'bar' } }, action)

      // Then
      expect(state).toEqual({ pending: {} })
    })
  })
})
