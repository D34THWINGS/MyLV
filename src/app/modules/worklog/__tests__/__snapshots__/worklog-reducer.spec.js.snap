// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worklog/reducer WORKLOG_FILL_MONTH should fill an entry for each working day of the month with given label 1`] = `
Object {
  "entries": Object {
    "0000-00-01-am": "world",
    "0000-00-01-pm": "world",
    "0000-00-02-am": "world",
    "0000-00-02-pm": "world",
    "0000-00-03-am": "world",
    "0000-00-03-pm": "world",
    "0000-00-04-am": "world",
    "0000-00-04-pm": "world",
    "0000-00-05-am": "world",
    "0000-00-05-pm": "world",
    "0000-00-08-am": "world",
    "0000-00-08-pm": "world",
    "0000-00-09-am": "world",
    "0000-00-09-pm": "world",
    "0000-00-10-am": "world",
    "0000-00-10-pm": "world",
    "0000-00-11-am": "world",
    "0000-00-11-pm": "world",
    "0000-00-15-am": "world",
    "0000-00-15-pm": "world",
    "0000-00-16-am": "world",
    "0000-00-16-pm": "world",
    "0000-00-17-am": "world",
    "0000-00-17-pm": "world",
    "0000-00-18-am": "world",
    "0000-00-18-pm": "world",
    "0000-00-19-am": "world",
    "0000-00-19-pm": "world",
    "0000-00-22-am": "world",
    "0000-00-22-pm": "world",
    "0000-00-23-am": "world",
    "0000-00-23-pm": "world",
    "0000-00-24-am": "world",
    "0000-00-24-pm": "world",
    "0000-00-25-am": "world",
    "0000-00-25-pm": "world",
    "0000-00-26-am": "world",
    "0000-00-26-pm": "world",
    "0000-00-29-am": "world",
    "0000-00-29-pm": "world",
    "0000-00-30-am": "world",
    "0000-00-30-pm": "world",
  },
}
`;

exports[`worklog/reducer WORKLOG_FILL_WEEK should fill an entry for each day of the week with given label 1`] = `
Object {
  "entries": Object {
    "foo-1-am": "world",
    "foo-1-pm": "world",
    "foo-2-am": "world",
    "foo-2-pm": "world",
    "foo-3-am": "world",
    "foo-3-pm": "world",
    "foo-4-am": "world",
    "foo-4-pm": "world",
    "foo-5-am": "world",
    "foo-5-pm": "world",
  },
}
`;
